src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,347 @@
+ */
+
+#pragma once
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Actions {
+
+inline constexpr size_t kActionNameMaxSize       = 128u;
+inline constexpr size_t kEndpointListNameMaxSize = 128u;
+
+class Delegate;
+
+struct GenericActionStruct : public Structs::ActionStruct::Type
+{
+    GenericActionStruct(){};
+
+    GenericActionStruct(uint16_t action, CharSpan label, ActionTypeEnum actionType, uint16_t epListID,
+                        BitMask<CommandBits> commands, ActionStateEnum actionState)
+    {
+        Set(action, label, actionType, epListID, commands, actionState);
+    }
+
+    GenericActionStruct(const GenericActionStruct & action) { *this = action; }
+
+    GenericActionStruct & operator=(const GenericActionStruct & action)
+    {
+        Set(action.actionID, action.name, action.type, action.endpointListID, action.supportedCommands, action.state);
+        return *this;
+    }
+
+    void Set(uint16_t action, CharSpan label, ActionTypeEnum actionType, uint16_t epListID, BitMask<CommandBits> commands,
+             ActionStateEnum actionState)
+    {
+        actionID          = action;
+        type              = actionType;
+        endpointListID    = epListID;
+        supportedCommands = commands;
+        state             = actionState;
+        memset(mActionNameBuffer, 0, sizeof(mActionNameBuffer));
+        if (label.size() > sizeof(mActionNameBuffer))
+        {
+            memcpy(mActionNameBuffer, label.data(), sizeof(mActionNameBuffer));
+            name = CharSpan(mActionNameBuffer, sizeof(mActionNameBuffer));
+        }
+        else
+        {
+            memcpy(mActionNameBuffer, label.data(), label.size());
+            name = CharSpan(mActionNameBuffer, label.size());
+        }
+    }
+
+private:
+    char mActionNameBuffer[kActionNameMaxSize];
+};
+
+struct GenericEndpointList : public Structs::EndpointListStruct::Type
+{
+    GenericEndpointList(){};
+
+    GenericEndpointList(uint16_t epListId, CharSpan label, EndpointListTypeEnum epListType,
+                        DataModel::List<const EndpointId> endpointList)
+    {
+        Set(epListId, label, epListType, endpointList);
+    }
+
+    GenericEndpointList(const GenericEndpointList & epList) { *this = epList; }
+
+    GenericEndpointList & operator=(const GenericEndpointList & epList)
+    {
+        Set(epList.endpointListID, epList.name, epList.type, epList.endpoints);
+        return *this;
+    }
+
+    void Set(uint16_t epListId, CharSpan label, EndpointListTypeEnum epListType, DataModel::List<const EndpointId> endpointList)
+    {
+        endpointListID = epListId;
+        type           = epListType;
+        endpoints      = endpointList;
```

Comment:
The "endpointList" needs to be copied into local storage too, no?  Otherwise we're just pointing to some externally-provided buffer....